app/controllers/protocols_controller.rb

```diff
@@ -658,194 +822,286 @@ def export
         }
-      }
+      end
     end
   end
 
   def edit_keywords_modal
     respond_to do |format|
-      format.json {
+      format.json do
         render json: {
           title: I18n.t('protocols.header.edit_keywords_modal.title',
                         protocol: escape_input(@protocol.name)),
-          html: render_to_string({
-            partial: "protocols/header/edit_keywords_modal_body.html.erb"
-          }),
+          html: render_to_string(partial: 'protocols/header/edit_keywords_modal_body.html.erb'),
           keywords: @protocol.team.protocol_keywords_list
         }
-      }
+      end
     end
   end
 
   def edit_authors_modal
     respond_to do |format|
-      format.json {
+      format.json do
         render json: {
           title: I18n.t('protocols.header.edit_authors_modal.title',
                         protocol: escape_input(@protocol.name)),
-          html: render_to_string({
-            partial: "protocols/header/edit_authors_modal_body.html.erb"
-          })
+          html: render_to_string(partial: 'protocols/header/edit_authors_modal_body.html.erb')
         }
-      }
+      end
     end
   end
 
   def edit_description_modal
     respond_to do |format|
-      format.json {
+      format.json do
         render json: {
           title: I18n.t('protocols.header.edit_description_modal.title',
                         protocol: escape_input(@protocol.name)),
-          html: render_to_string({
-            partial: "protocols/header/edit_description_modal_body.html.erb"
-          })
+          html: render_to_string(partial: 'protocols/header/edit_description_modal_body.html.erb')
         }
-      }
+      end
     end
   end
 
   private
 
+  def protocolsio_step_description_populate(result, iterating_key, pos2)
+    if !iterating_key['data'].nil? && iterating_key['data'] != '' &&
+       result['steps'][pos2.to_s]['description']
+      append = '<br>' + iterating_key['data'] + '<br>'
+      result['steps'][pos2.to_s]['description'] << append
+    elsif !result['steps'][pos2.to_s]['description']
+      append = '<br>' + iterating_key['data'] + '<br>'
+      result['steps'][pos2.to_s]['description'] = append
+    else
+      result['steps'][pos2.to_s]['description'] = 'Description missing'
+    end
+    result
+  end
+
+  def protocolsio_step_title_populate(result, iterating_key, pos2)
+    result['steps'][pos2.to_s]['name'] =
+      if !iterating_key['data'].nil? && iterating_key['data'] != ''
+        iterating_key['data']
+      else
+        'Step'
+      end
+    result
+  end
+
+  def protocolsio_step_expected_result_populate(result, iterating_key, pos2)
+    if !iterating_key['data'].nil? && iterating_key['data'] != ''
+      append = '<br><strong>Expected result: </strong>'
+      + iterating_key['data'] + '<br>'
+      result['steps'][pos2.to_s]['description'] << append
+    end
+    result
+  end
+
+  def protocolsio_step_software_package_populate(result, iterating_key, pos2)
```

Comment:
It's better that you pass 'iterating_key['source_data']' parameter to the method, because you don't need the other data from 'iterating_key'.
And if the other passed parameter would be 'result['steps'][pos.to_s]', then you need just one param instead of two (result, pos2).